docs/tutorials/developer-liftoff-rust/level-1/1.1-rust-lvl1.mdx

```diff
@@ -0,0 +1,227 @@
+
+A canister's `init` entry point should match the initialization parameters in the Candid interface. It must have no return value. Each canister can only have one `init` entry point.
+
+```rust
+#[ic_cdk::init]
+fn init_function() {
+    // ...
+}
+```
+
+Refer to the [`canister_init` specification](https://internetcomputer.org/docs/current/references/ic-interface-spec/#system-api-init) for more information.
+
+### `query`
+
+A `query` endpoint provides a callable method that only returns information. `query` endpoints cannot alter the state of a canister.
+
+```rust
+#[ic_cdk::query]
+fn query_function() {
+    // ...
+}
+```
+
+If you specify a name for the method, the canister will expose the method on the public interface with that name. Otherwise, the name of the function will be used:
+
+```rust
+#[ic_cdk::query(name = "method_name")]
+fn query_function() {
+    // ...
+}
+```
+
+If a `query` endpoint must have a prerequisite executed before it can be called, you can define a `guard` function that must be executed before the query function can be executed. If the `guard` function returns an error, the `query` function does not execute.
+
+```rust
+fn guard_function() -> Result<(), String> {
+    // ...
+}
+
+#[ic_cdk::query(guard = "guard_function")]
+fn query_function() {
+    // ...
+}
+```
+
+### `update`
+
+An `update` entry point provides a callable method that can make changes to the canister's state. `update` methods may return a response or they may return an empty type.
+
+
+```rust
+#[ic_cdk::update]
+fn update_function() {
+    // ...
+}
+```
+
+`update` methods can be named or configured to use `guard` functions in the same ways that `query` methods can.
+
+## Candid interface files
+
+Candid is an interface description file (IDL) used to describe a canister's public methods. Each Candid file defines a service, then adds the canister's public methods and the data types that each method accepts and returns.
+
+Here is an example that defines the public method `greet`, which accepts type `text` and returns type `text` using a `query` call.
+
+```did
+service : {
+    "greet" : (text) -> (text) query;
+};
+```
+
+Candid interface description files (`.did`) for Rust canisters must be handwritten or manually exported by first defining the `export_candid!` macro, which exposes a method called `get_candid_pointer`, which `dfx` uses to extract the canister's Candid definition. Then you must use the [`candid-extractor`](https://crates.io/crates/candid-extractor) tool to extract the canister's Candid file. This macro must be called once at the end of your canister's code outside of any query or update definitions.
```

Comment:
This is sort of misleading. dfx doesn't extract the candid, this is purely the province of candid-extractor. dfx asks for a file, and candid-extractor can produce one instead of you writing one. Otherwise you might get the impression that this is automatic.